javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll

```diff
@@ -69,7 +69,28 @@ abstract class AtmConfig extends string {
    * Holds if the candidate sink `candidateSink` predicted by the machine learning model should be
    * an effective sink, i.e. one considered as a possible sink of flow in the boosted query.
    */
-  predicate isEffectiveSink(JS::DataFlow::Node candidateSink) { none() }
+  predicate isEffectiveSink(JS::DataFlow::Node candidateSink) {
+    not exists(this.getAReasonSinkExcluded(candidateSink))
+  }
+
+  final EndpointCharacteristics::EndpointCharacteristic getAReasonSinkExcluded(
+    JS::DataFlow::Node candidateSink
+  ) {
+    // An endpoint is an effective sink if it has neither standard endpoint filter characteristics nor endpoint filter
+    // characteristics that are specific to this sink type.
+    // TODO: Experiment with excluding all endpoints that have a medium- or high-confidence characteristic that implies
+    // they're not sinks for this sink type (or not sinks for any sink type), not just the EndpointFilterCharacteristics.
+    exists(EndpointCharacteristics::StandardEndpointFilterCharacteristic standardFilter |
```

Comment:
Again, see the TODO comment right above this. I don't know if EndpointFilterCharacteristics will even exist in the long run (I hope not). They are a way of replicating the existing logic with the new framework. In the long run all selection (training and inference) should ideally depend only on the implications of the characteristics, and not on their class structure. If being an EndpointFilterCharacteristic doesn't carry any particular implications, then the selection code shouldn't know or care that this class exists.

Reply:
Again see my TODO comment directly above this. Same thing here as all the other places. I believe that's the right approach, but it needs to be verified experimentally.

Reply:
A TODO would be the appropriate choice if the fix would take a lot of work; but the fix here is a matter of minutes.

If you're currently juggling too many things at the same time, I'd be more than happy to help out by adding a small commit first thing tomorrow morning! :) Just let me know in that case.

Reply:
Making this change requires end-to-end testing to ensure it doesn't negatively impact our metrics. That's far from trivial, though it will be easier once we have the MVP of the orchestrator. Like we agreed, I'm first making the transition to the new framework _without_ impacting the data we produced, and then we can start doing all the experiments we want to improve the data.

Reply:
> Making this change requires end-to-end testing to ensure it doesn't negatively impact our metrics.

That's not true; I just redid what I did yesterday locally. Using the clause:

```
exists(EndpointCharacteristics::EndpointCharacteristic filter, float c |
      filter.getEndpoints(candidateSink) and
      c >= any(EndpointCharacteristics::EndpointCharacteristic ch).mediumConfidence() and
      filter.getImplications(any(NegativeType n), true, c) and
      result = standardFilter
    )
```

filters exactly the same endpoints as your clause, but in a way that is more modular. If it would've changed the data selection, I'd not have (knowingly) suggested a change.

> Incidentally, it wouldn't be c >= any(EndpointCharacteristic c).maximalConfidence(): endpoint filters have medium confidence, not maximal confidence.

I know, I had mentioned this in the first comment of this thread.

Reply:
Endpoint filters are _not_ maximal confidence -- quite the opposite, they're lower confidence than NotASink. AIUI, your commit should have added lots of negative examples to the training set that weren't there before -- I'm not sure why it didn't fail `endpoint_large_scale/ExtractEndpointDataTraining.qlref`. At the moment there isn't a test that verifies the selection of endpoints for inference (see https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2135), so that needs to be verified via a DCA experiment. 

In any case, I could definitely achieve something like this by carefully setting and then filtering on confidence levels, to ensure that in practice the training set chooses only the historical NotASink endpoints and inference runs only on the historical effective sink endpoints. I feel this would hide the fact that these current choices are fragile and often arbitrary, though. I prefer to be explicit about the fact that we have some arbitrary selection going on at the moment, by explicitly using subclasses for selection, and then change that ugliness when we're truly rationalizing the selection by setting proper confidence levels and using them consistently.

Reply:
TL;DR 🤝 : I think we may have found a compromise.

At this point, I want to explain why I care about this so much; it may seem like an irrelevant detail, but to me it's far from it.

The framework we're building is based on expressing our believes about classes of endpoints in terms of implications, which are labels involving a sink type, a boolean, and a confidence. This is a really simple framework (I mean 'simple' in a good way). However, if this framework would turn out to be TOO simple to express what we need to express, we're painting ourselves in a corner: That's why I react strongly to all cases where selection of endpoints is not based on the implications, but on the specific characteristic class. in these cases, the implications framework was not sufficient to express what we need to express (or it was too hard). Saying that these situations will all be resolved later on in the process (as the TODOs do) can't be done with certainty. In order to remove those cases, we'll need experiments to show that resolving the situations does not make the models worse. If the odd legacy rules happen to be weird, but work in practice, we'd be stuck with that strange logic for a long time, or forever.

As we have decided to go with a process that splits the refactoring into many small PRs, it's extra important to address these concerns earlier, rather than later — later, the resistance to undo so much work (in the worst case) would be huge.

> Endpoint filters are not maximal confidence -- quite the opposite, they're lower confidence than NotASink

I'm totally with you there - I don't care about using this specific value at all, any value that'll reproduce the current behaviour will suffice. *BUT*, I'd say that this assignment of values looks so strange is a *feature* of the new framework: it makes this weirdly high confidence setting stick out like a sore thumb.

> In any case, I could definitely achieve something like this by carefully setting and then filtering on confidence levels, to ensure that in practice the training set chooses only the historical NotASink endpoints and inference runs only on the historical effective sink endpoints. I feel this would hide the fact that these current choices are fragile and often arbitrary, though.

These choices, on SOME level will always be arbitrary, as we only merge settings that are backed by experimental evidence. That being said, we both are committed to finding better values for them in the future, which means it will definitely happen.

My current worry, though, is about expressivity of the framework, and in the future it may be too late to make changes.

If you'd be willing to use any values that make it all work, we'd have a compromise that I could feel almost relaxed about (you know me, I'm never *really* relaxed, so that's as good as it gets ^^), merge this PR and move on to the next.

If you want extra assurance that adapting the confidence levels gets tackled later: I can make an issue and assign it to myself so we don't lose track of that promise. Or if you'd rather own that, we can assign it to you - but I'm offering to do the work, as I think this might get us to the handshake.

Reply:
OK, thank you for clarifying -- I have a much better understanding of your concerns now 🙏 I suspect we'll still have some back-and-forth on this, but we're getting closer to understanding one another's concerns, which is a win!

My guess is we won't be _entirely_ able to replace all the current logic using only confidence levels, because at some point we'll hit something where logically we think a characteristic has confidence X but in practice, experimentally we get bad results when we do that. And that's OK -- a rule is still useful even when it has a small number of exceptions: the key is to make those exceptions as clear and visible as possible, and make the entire codebase easy to understand and maintain. In my mind there's no question that this new framework accomplishes that, even if we won't be able to remove every single edge case we're currently reproducing from the old code. It'll be very easy to see what the remaining edge cases are and why they're there, which was _not_ easy to see with the old framework. The edge cases will be encoded in exactly two places: edge cases for the training set will be encoded explicitly in the [predicate that selects endpoints for training](https://github.com/github/codeql/blob/3fbe089f65c291030389ef41ba995a5506783346/javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll#L53), and edge cases for the inference set will be encoded explicitly in the predicate that selects endpoints for inference (which is this one). Setting weird confidence levels hides these exceptions. Calling the exceptions out directly in the extraction predicate makes them explicit and visible.

Reply:
I created [this commit](https://github.com/github/codeql/commit/03b8e649f1528905ed1709ce4581e9368337cccf) in a separate branch to show how easily we can replicate the current selection using only confidence levels. This is exactly what I think we want to do long-term, _except for_ line 86: `confidence < filter.highConfidence() and`. Logically it makes no sense to score endpoints that we are highly certain are _not_ sinks at inference time. The reason this line is needed is that there are currently three high-confidence (NotASink) characteristics that are shared by many TP sinks: 
- JQueryArgument (Xss sinks)
- FileSystemAccess (tainted path sinks)
- DatabaseAccess (NoSQL / SQL sinks)

When I delete that line, `endpoint_large_scale/FilteredTruePositives.ql` fails, with lots of newly-filtered TPs, which could well impact ATM metrics. Of course, `endpoint_large_scale/FilteredTruePositives.ql` already has many exceptions hard-coded in, which means some of the current endpoint filters also apply to a decent number of TPs. I don't know whether they were added to speed up inference or to improve ATM metrics, but I would hope we could surface more TPs by removing all these endpoint filters. We should certainly consider removing `JQueryArgument`, `FileSystemAccess`, and `DatabaseAccess` from the set of high-confidence characteristics, because they may be adding lots of bad negative examples to the training set.

Merging this commit would hide all these issues, and thereby reduce our ability to make adjustments that might improve ATM metrics considerably. At the moment, we have a set of decisions that were made for unknown reasons at some point in history. Some of them might indeed be critical to reaching good metrics and/or good runtime, but others are likely doing more harm than good. Only experimentation will tell us which of these decisions we can undo and which ones we have to keep. Of the ones we keep, preferably we can encode them by properly adjusting confidence levels, but there may be some corner cases where, for whatever reason, we want to, say include some type of endpoint in the negative training set without excluding it as a candidate for scoring 🤷

Reply:
Having said all that, if you feel very strongly about it, I would be willing to merge [this commit](https://github.com/github/codeql/commit/03b8e649f1528905ed1709ce4581e9368337cccf) (with some updates to the documentation), because it still upholds the rule that edge cases are encoded only in the endpoint extraction queries. In that sense, it still makes it easy to find all the edge cases, keeping them explicit and visible. What it doesn't preserve is our ability to find all edge cases simply by making all EndpointCharacteristic sub-classes private, but if you feel strongly about this that's a price I'm willing to pay 🤷‍♂️

Reply:
OK, then we have a deal! Yay! 🥳 This makes me happy to hear 👍

I'm still a bit concerned about how the confidences framework will work out in the end, but I thought lots about it last night and there's a plan B that should cover us in the very very worst case. So I think that merging with this commit is the reasonable thing to do.